File: KmpDemo/iosApp/iosApp/SwApiViewModel.swift
import Foundation
import Common

@MainActor
class SwApiViewModel: ObservableObject {
    @Published var state = SwApiUiState.idle

    var characterCount: Int {
        return Int(swapiService.getCharacterCount())
    }
    
    func getCharacter(_ characterId: String) -> Character? {
        return swapiService.getCharacterById(characterId: characterId)
    }
    
    func characterFailedToFetch(_ characterId: String) -> Bool {
        return swapiService.characterFailedToFetch(characterId: characterId)
    }
    
    private let swapiService = SwApiService()
    
    init() {
        state = SwApiUiState.initializing
        swapiService.loadInitialData { result, error in
            if result != nil && error == nil  {
                self.state = .characterLoaded
            } else {
                self.state = .initializationError
            }
        }
    }
    
    func loadCharacter(_ characterId: String) {
        if getCharacter(characterId) != nil {
            return
        }
        if characterFailedToFetch(characterId) {
            return
        }
        
        if !state.isBusy {
            state = .fetchingNewCharacter(characterId)
            swapiService.loadCharacterById(characterId: characterId) { result, _ in
                self.state = .characterLoaded
            }
        }
    }
    
}

enum SwApiUiState {
    case idle
    case initializing
    case fetchingNewCharacter(String)
    case characterLoaded
    case initializationError
    
    var isBusy: Bool {
        switch self {
        case .fetchingNewCharacter: return true
        default: return false
        }
    }
}

File: KmpDemo/androidApp/src/main/java/com/kalalaucantrell/kmpdemo/android/SwApiViewModel.kt
package com.kalalaucantrell.kmpdemo.android

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.kalalaucantrell.kmpdemo.SwApiService
import com.kalalaucantrell.kmpdemo.android.SwApiUiState.*
import com.kalalaucantrell.kmpdemo.models.Character
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.launch

class SwApiViewModel : ViewModel() {
    val uiState: MutableStateFlow<SwApiUiState> = MutableStateFlow(Idle())

    val characterCount: Int
        get() = swapiService.getCharacterCount()

    fun getCharacter(characterId: String): Character? {
        return swapiService.getCharacterById(characterId)
    }

    private fun characterFailedToFetch(characterId: String): Boolean {
        return swapiService.characterFailedToFetch(characterId)
    }

    private val swapiService = SwApiService()

    init {
        loadInitialData()
    }

    private fun loadInitialData() {
        uiState.value = Initializing()
        viewModelScope.launch {
            try {
                swapiService.loadInitialData()
                uiState.value = CharacterLoaded()
            } catch (error: Throwable) {
                uiState.value = InitializationError()
            }
        }
    }

    fun loadCharacter(characterId: String) {
        if (!uiState.value.isBusy) {
            if (getCharacter(characterId) != null) {
                return
            }
            if (characterFailedToFetch(characterId)) {
                return
            }
            uiState.value = FetchingNewCharacter(characterId)
            viewModelScope.launch {
                swapiService.loadCharacterById(characterId)
                uiState.value = CharacterLoaded()
            }
        }
    }
}

sealed class SwApiUiState {
    data class Idle(val idle: Boolean = true) : SwApiUiState()
    data class Initializing(val loading: Boolean = true) : SwApiUiState()
    data class FetchingNewCharacter(val characterId: String) : SwApiUiState()
    data class CharacterLoaded(val loaded: Boolean = true) : SwApiUiState()
    data class InitializationError(val error: Boolean = true) : SwApiUiState()

    val isBusy: Boolean
        get() = when (this) {
            is FetchingNewCharacter -> true
            else -> false
        }
}
